Add a hasTraps() check in B3::Value::effects()

kmiller68 opened the pull request:
#### a649419b8cf4bc2d1b95dfc5c2ba17bf4310de8c
<pre>
Add a hasTraps() check in B3::Value::effects()
<a href="https://bugs.webkit.org/show_bug.cgi?id=273648">https://bugs.webkit.org/show_bug.cgi?id=273648</a>
<a href="https://rdar.apple.com/127458113">rdar://127458113</a>

Reviewed by Yusuke Suzuki.

Some profiling of reduceStrength showed we spent a number of samples
in B3::Value::effects() checking traps(). But most B3 Kinds don&apos;t
ever have traps(). Since we just switched on the Kind we can use
`hasTraps()` hoist the traps() check into the cases that actually
use it.

* Source/JavaScriptCore/b3/B3EliminateDeadCode.cpp:
(JSC::B3::eliminateDeadCodeImpl):
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):

Canonical link: <a href="https://commits.webkit.org/278338@main">https://commits.webkit.org/278338@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ab61b30e3596819b76c739d9ecc0cef582de04a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50107 "Passed style check") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29398 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Running checkout-pull-request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2400 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Running checkout-pull-request") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/796 "Build is in progress. Recent messages:Running configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35564 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/352 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Running checkout-pull-request") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40896 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52205 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/35564 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/2400 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Running checkout-pull-request") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21987 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Running run-api-tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/35564 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/2400 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Running checkout-pull-request") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8484 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/43433 "Build is in progress. Recent messages:OS: Ventura (13.6.6), Xcode: 14.3") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/35564 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/2400 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Running checkout-pull-request") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54946 "Build is in progress. Recent messages:Printed configuration") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/49605 "Build is in progress. Recent messages:OS: Ventura (13.6.6), Xcode: 14.3; Running apply-patch; Checked out pull request; Running compile-jsc") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25203 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/352 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Running checkout-pull-request") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/54946 "Build is in progress. Recent messages:Printed configuration") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/2400 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Running checkout-pull-request") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/54946 "Build is in progress. Recent messages:Printed configuration") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27327 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/57084 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7261 "Built successfully and passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26195 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11726 "Build is in progress. Recent messages:Running apply-patch; Checked out pull request; Running jscore-test") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -3652,9 +3654,12 @@ class ReduceStrength {
 
 bool reduceStrength(Procedure& proc)
 {
+    WTFBeginSignpostAlways(nullptr, JSCJITCompiler);
```
Let's put behind `Options::useCompilerSignpost()`.

kmiller68 replied:
Whoops, didn't mean to upload that... Will remove.

Constellation commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -3652,9 +3654,12 @@ class ReduceStrength {
     ReduceStrength reduceStrength(proc);
-    return reduceStrength.run();
+    bool result = reduceStrength.run();
+    WTFEndSignpostAlways(nullptr, JSCJITCompiler);
```
Ditto.

Constellation reviewed (approved):
r=me with comments.

kmiller68 committed:
Add a hasTraps() check in B3::Value::effects()
https://bugs.webkit.org/show_bug.cgi?id=273648
rdar://127458113

Reviewed by Yusuke Suzuki.

Some profiling of reduceStrength showed we spent a number of samples
in B3::Value::effects() checking traps(). But most B3 Kinds don't
ever have traps(). Since we just switched on the Kind we can use
`hasTraps()` hoist the traps() check into the cases that actually
use it.

* Source/JavaScriptCore/b3/B3EliminateDeadCode.cpp:
(JSC::B3::eliminateDeadCodeImpl):
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):

Canonical link: https://commits.webkit.org/278338@main
